lib/libxdp/libxdp.c

```diff
@@ -1956,16 +1957,36 @@ static int xdp_multiprog__check_compat(struct xdp_multiprog *mp)
 		goto out;
 	}
 
-	lfd = bpf_raw_tracepoint_open(NULL, test_prog->prog_fd);
-	if (lfd < 0) {
+	test_prog->link_fd = bpf_raw_tracepoint_open(NULL, test_prog->prog_fd);
+	if (test_prog->link_fd < 0) {
```

Comment:
Can't you just manually pin the lfd at some other location instead of doing the whole "replace internal pointer then call __pin()" dance? That should keep you from having to do the timeout as well, shouldn't it?

Reply:
You mean calling ```bpf_obj_pin``` but with a different path? When I tried that and it didn't help. But I looked through the code a little bit and the busy error is coming from ```bpf_trampoline_link_prog``` caused by ```bpf_trampoline_lookup``` returning trampoline that's not unlinked yet and still has a pointer to the old program. And the trampoline is found using combination of ```main_prog->prog_id``` and ```btf_id```, so it doesn't depend on the pin location. But changing ```bpf_program__set_attach_target``` to different target instead helped. So to fix this and get rid of the timout maybe just instead of ```prog0``` use ```xdp_dispatcher``` (since that one is not getting pinned). 
And I avoided using ```bpf_obj_pin``` directly because I would also have to add a bunch of other parts of the pinning function, but in retrospect it wouldn't be that bad.

Reply:
Ah, I see. Hmm, I don't think using xdp_dispatcher for the freplace target is a good idea if the problem is that it's not being detached in time. Because that means that it can still be attached by the time we're actually attaching the dispatcher to the interface, which would be bad.

Instead, let's just add a separate target only for this purpose to the dispatcher file. I.e., in xdp-dispatcher.c.in, after loop that generates the prog* functions, add a compat_test function:

```diff
diff --git a/lib/libxdp/xdp-dispatcher.c.in b/lib/libxdp/xdp-dispatcher.c.in
index a601e23d5b32..dc9a858e9f3f 100644
--- a/lib/libxdp/xdp-dispatcher.c.in
+++ b/lib/libxdp/xdp-dispatcher.c.in
@@ -40,6 +40,15 @@ int format(`prog%d', i)(struct xdp_md *ctx) {
 }
 ')
 
+__attribute__ ((noinline))
+int compat_test(struct xdp_md *ctx) {
+        volatile int ret = XDP_DISPATCHER_RETVAL;
+
+        if (!ctx)
+          return XDP_ABORTED;
+        return ret;
+}
+
 SEC("xdp/dispatcher")
 int xdp_dispatcher(struct xdp_md *ctx)
 {
@@ -53,6 +62,12 @@ forloop(`i', `0', NUM_PROGS,
         if (!((1U << ret) & conf.chain_call_actions[i]))
                 return ret;
 ')
+        /* keep a reference to the compat_test() function so we can use it
+         * as an freplace target in xdp_multiprog__check_compat() in libxdp
+         */
+        if (num_progs_enabled < incr(NUM_PROGS))
+                goto out;
+        ret = compat_test(ctx);
 out:
         return XDP_PASS;
 }
```

and then use that as the target...